src/System.Management.Automation/engine/serialization.cs

```diff
@@ -4039,38 +4053,78 @@ private object ReadDictionary(ContainerType ct)
 
                     object value = ReadOneObject();
 
-                    // On the first collision, copy the hash table to one that uses the default comparer.
-                    if (table.ContainsKey(key) && (keyClashFoundIteration == 0))
+                    if (!isOrdered)
                     {
-                        keyClashFoundIteration++;
-                        Hashtable newHashTable = new Hashtable();
-                        foreach (DictionaryEntry entry in table)
+                        // On the first collision, copy the hash table to one that uses the default comparer.
+                        if (table.ContainsKey(key) && (keyClashFoundIteration == 0))
                         {
-                            newHashTable.Add(entry.Key, entry.Value);
+                            keyClashFoundIteration++;
+                            Hashtable newHashTable = new Hashtable();
```

Comment:
```suggestion
                            Hashtable newHashTable = new Hashtable(table.Count);
```

Below too.

Reply:
Changed in both cases (again, was in the original code).